edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/ConnectionHandler.cs

```diff
@@ -152,6 +152,22 @@ async Task ReconcileConnectionsAsync(HashSet<IIdentity> updatedIdentities)
 
                 await this.RemoveConnectionsAsync(identitiesRemoved);
                 await this.AddConnectionsAsync(identitiesAdded);
+
+                // When a device connects indirectly, then disconnects and connects directly again, the following happens:
+                // As we don't have "indirect disconnection", the device proxy instance lingers for a while before gets deleted
+                // by idle-time. Because of that it may show up in the "knownIdentities" list, and the way added connections are
+                // calculated above, it will not be added as a "direct connection" but keeps existing as "indirect connection".
+                // As a result, the device proxy will use mqtt topics for indirect connections which will not be subscribed
```

Comment:
I am a little confused by this line: `it will not be added as a "direct connection" but keeps existing as "indirect connection"`

As I thought the issue was when having setup like L5->L4->L3
testing with L4 as gateway for twin update to L3 would be a direct connection
testing with L5 as gateway for twin update to L3 would be an indirection connection
So, testing with L4 first and then L5 would be that L5 as gateway was trying to use old 'direct' connection but should use 'indirect'. Is this understanding correct? As from the comment at least it feels like the opposite is being said. Not saying this wrong, just want to understand if our initial analysis the other day had the terms flipped or something :)

Reply:
If we have L5->L4->L3 and L3 connects to L4, then L5 sees L3 as indirect (because l5 never received a connection event from L3 but saw L3 related communication, so it assumed that L3 connected somewhere else and now it is talking).
When L3 disconnects from L4, L5 will not know about the disconnection and keeps its state as L3 is still connected "indirectly". Now if L3 connects to L5, then L5 receives a connection event, but as it has L3 represented in its internal state, it keeps it that way with the wrong 'indirect' flag.

If L3 was connected L5 first (before it connected L4), then L5 receives a connected event (which is not really important now), and when it goes away, it receives a disconnected event (which is the important part). So now L5 knows that L3 went away and it updates its state (it deletes the L3 related internal objects). So when L3 comes back and connects to L4, then start talking (e.g. requests a twin), L5 will learn about L3 transmitting the request and it creates L3 related objects with the "indirect" flag. That is why it was not a problem if L3 comes to L5 first, then leaves and comes back through L4. But the other way around this was a problem. And when L3 comes to L4 first, L5 will have and "indirect connection" and when L3 connects to L5 a bit later, this "indirect connection" is kept (but should be deleted and recreated)